### ISSUE TITLE
Failed to load config "airbnb-base" to extend from. Referenced from: /app/.eslintrc.js

### ISSUE THREAD

**Issue by r-wells** [airbnb/javascript#2481]
As the title suggests, I keep getting this error above when trying to access my app on a **production** url in Heroku:

```
Failed to load config "airbnb-base" to extend from.
Referenced from: /app/.eslintrc.js
```

I **do not** have this issue on my development server. Everything loads and works how I need it to.

Which leads me to believe there are certain configs I need to setup on Heroku to deploy it correctly.

That being said, none of the solutions to the other similar issues I've researched have done anything for me.

My .eslintrc.js file is below:

```
function extractRules(rules) {
  const outputRules = {};

  Object.keys(rules).forEach((ruleName) => {
    const rule = rules[ruleName];

    if (!(rule.justification && rule.justification.length > 0)) {
      throw Error(
        `All linting rules must include a justification.
            Please specify one for ${ruleName}.`,
      );
    }

    outputRules[ruleName] = rule.value;
  });

  return outputRules;
}

module.exports = {
  root: true,
  extends: [
    'react-app',
    'react-app/jest',
    'airbnb-base',
    'airbnb/hooks',
  ],
  env: {
    browser: true,
    jest: true,
  },
  plugins: [
    'jest-dom',
    'react',
    'testing-library',
  ],
  rules: extractRules({
    'import/extensions': {
      value: ['error', 'ignorePackages', {
        js: 'never', jsx: 'never', ts: 'never', tsx: 'never',
      }],
      justification: `AirBnB misconfiguration complains about missing file
          extension even though there is nothing wrong. This fixes it.
          See https://stackoverflow.com/a/59268871 for more information.`,
    },
    'react/jsx-filename-extension': {
      value: ['error', { extensions: ['.jsx', '.tsx'] }],
      justification: `Allow '.tsx' files to contain JSX.`,
    },

    // other global rules
    'arrow-parens': {
      value: ['error', 'always'],
      justification: `Require parenthesis around arrow function arguments for
          increased readability.`,
    },
    'camelcase': {
      value: 'off',
      justification: `Our API responds in snake case, so sometimes we can't
          adhere to camelcase only.`,
    },
    'import/no-unresolved': {
      value: 'warn',
      justification: `Warn developer when a path cannot be resolved.`,
    },
    'import/no-extraneous-dependencies': {
      value: ['error', {
        devDependencies: [
          '**/*.test.jsx',
          '**/*.test.tsx',
          '**/*.spec.jsx',
          '**/*.spec.tsx',
          'src/setupTests.js',
          '/src/utils/testing*',
        ],
      }],
      justification: `Importing dev dependencies is expected in tests, and
          so we instruct this rule to ignore testing files.
          See https://bit.ly/3eNmzW5 for more information.`,
    },
    'key-spacing': {
      // TODO: Switch over rule.
      // value : ['warn', {
      //   multiLine : { beforeColon: true, afterColon: true },
      //   align     : { beforeColon: true, afterColon: true, on: 'colon' },
      // }],
      value: 'off',
      justification: `This aligns the colon on object literals for increased
          readibility. This is disabled at present to prevent changes to too
          many files.`,
    },
    'no-plusplus': {
      value: ['error', { allowForLoopAfterthoughts: true }],
      justification: `It is best practice not to allow '++', but for-loops
          are conventionally written with them, and so we allow them here.`,
    },
    'no-unused-vars': {
      value: ['warn', { argsIgnorePattern: '^_' }],
      justification: `In general, unused variables should be deleted, but in
          instances where the developer would like to explicitly include an
          usused variable, they can deliberately acknowledge that the variable is
          meant to be used by prefixing it with an underscore. This is
          particularly useful when destructuring arrays.`,
    },
    'no-warning-comments': {
      // TODO: Enable this in the future once TypeScript conversion is complete.
      // value         : ['warn', { terms: ['bugbug', 'fixme', 'todo', 'xxx'] }],
      value: 'off',
      justification: `We do use warning comments to flag technical debt. This
          allows these flags to be raised to increase awareness that it exists.`,
    },
    'object-curly-newline': {
      value: 'off',
      justification: `Imports are cleaner to read and easier to reorder if
          they are in line with the object curly braces. This disables forcing
          them to be on new lines except where it would cause max-len to be
          violated.`,
    },
    'quote-props': {
      value: ['error', 'consistent-as-needed'],
      justification: `Consistency and readability. Seeing mixed quoting of
          object properties can be difficult to read.`,
    },
    'quotes': {
      value: ['error', 'single', {
        allowTemplateLiterals: true,
        avoidEscape: true,
      }],
      justification: `Single quotes are a team convention, and this allows for
          the usage of template literals and double quotes to avoid escaping.`,
    },
    'react/no-multi-comp': {
      // TODO: This should eventually become an error.
      value: ['warn', { ignoreStateless: true }],
      justification: `React best practice dictates one component per file for
          readability and reusability.`,
    },
    'react/prop-types': {
      // TODO: Delete this rule once TypeScript conversion is complete.
      value: 'off',
      justification: `This project does not use PropTypes and is transitioning
          to Typescript.`,
    },
    'react/require-default-props': {
      // TODO: Delete this rule once TypeScript conversion is complete.
      value: 'off',
      justification: `This project does not use PropTypes and is transitioning
          to Typescript.`,
    },
    'react-hooks/exhaustive-deps': {
      value: 'warn',
      justification: `We still want this to be flagged, but not to error.`,
    },
    'semi-style': {
      value: 'off',
      justification: `There are instances where developers might want to place
          semicolons in specific places for readability.`,
    },
    'space-before-function-paren': {
      value: ['error', {
        anonymous: 'never',
        named: 'never',
        asyncArrow: 'always',
      }],
      justification: `Improves readability, especially with TypeScript where
          types and generics might complicate a function signature.`,
    },
  }),
  overrides: [
    { // Typescript-specific config
      files: ['*.ts', '*.tsx'],
      parserOptions: {
        project: [
          './tsconfig.json',
        ],
      },
      extends: [
        'plugin:@typescript-eslint/recommended',
      ],
      rules: extractRules({
        '@typescript-eslint/no-empty-function': {
          value: 'off',
          justification: `Allowing no-ops are a team convention.`,
        },
        '@typescript-eslint/no-empty-interface': {
          value: 'off',
          justification: `Allows empty interfaces (primarily for 'react-table'
              types).`,
        },
        '@typescript-eslint/no-shadow': {
          value: ['error'],
          justification: `Reduces confusion with having to track variable name
              across scope.`,
        },
        'no-multi-spaces': {
          // TODO: Enable globally once TypeScript conversion is complete.
          value: 'off',
          justification: `Sometimes, multiple spaces can be better for
              redability, especially when defining types. This rule is enabled
              for JavaScript files, but not for TypeScript files.`,
        },
        'no-shadow': {
          value: 'off',
          justification: `Disable JS rule in favor of TS rule.`,
        },
        'no-use-before-define': {
          value: 'off',
          justification: `Disable so we can import react, rule has been known to report incorrect errors.`,
        },
      }),
    },
    { // JSX-specific config
      files: ['*.jsx', '*.tsx'],
      rules: extractRules({
        'jsx-a11y/anchor-is-valid': {
          // TODO: This should be converted to an error eventually.
          value: 'off',
          justification: `Preexisting team convention.`,
        },
        'jsx-a11y/click-events-have-key-events': {
          // TODO: This should be converted to an error eventually.
          value: 'off',
          justification: `Preexisting team convention.`,
        },
        'jsx-a11y/no-static-element-interactions': {
          // TODO: This should be converted to an error eventually.
          value: 'off',
          justification: `Preexisting team convention.`,
        },
        'react/jsx-max-props-per-line': {
          value: ['warn', { maximum: 1 }],
          justification: `Keeping each prop on its own line increases
              readability for JSX files.`,
        },
        'react/jsx-props-no-spreading': {
          value: 'off',
          justification: `Preexisting team convention.`,
        },
        'react/react-in-jsx-scope': {
          value: 'off',
          justification: `React 17, which we are using, does not require React
              to be in scope when using JSX.`,
        },
      }),
    },
    { // Test-specific config
      files: [
        '**/*.spec.jsx',
        '**/*.spec.tsx',
        '**/*.test.jsx',
        '**/*.test.tsx',
      ],
      extends: [
        'plugin:testing-library/react',
        'plugin:jest-dom/recommended',
      ],
    },
  ],
  settings: {
    'import/resolver': {
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx'],
      },
    },
  },
};

```

I'm really confused here. One thing I'm assuming is that since [Heroku removes devDependencies](https://www.digitalocean.com/community/questions/how-do-you-use-node-js-devdependencies-in-the-app-platform-builds) after the build is complete, the `.eslintrc.js` file doesn't know what to reference.

But this doesn't explain why I'm not having this problem with my other react app deployed on Heroku, with essentially the same exact `.eslintrc.js` configuration.

Also this is my `package.json` for reference:

```
{
  "name": "react-blueprint",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@auth0/auth0-react": "^1.8.0",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^11.2.7",
    "@testing-library/user-event": "^12.8.3",
    "@types/jest": "^26.0.24",
    "@types/node": "^12.20.28",
    "@types/react": "^17.0.27",
    "@types/react-dom": "^17.0.9",
    "axios": "^0.22.0",
    "deserialize-json-api": "^1.4.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^5.3.0",
    "react-scripts": "4.0.3",
    "typescript": "^4.4.3",
    "web-vitals": "^1.1.2"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "lint": "eslint . --ext .js,.jsx,.ts,.tsx"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@babel/core": "^7.13.0",
    "@types/react-router-dom": "^5.3.1",
    "babel-eslint": "^10.1.0",
    "eslint": "^7.32.0",
    "eslint-config-airbnb": "^18.2.1",
    "eslint-config-airbnb-base": "^14.2.1",
    "eslint-config-airbnb-typescript": "^14.0.1",
    "eslint-plugin-import": "^2.25.2",
    "eslint-plugin-jest-dom": "^3.9.2",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-react": "^7.26.1",
    "eslint-plugin-react-hooks": "^4.0.8",
    "eslint-plugin-testing-library": "^4.12.4"
  }
}

```

Would appreciate any help. Again, this is not an issue on a development server, but on a production build in Heroku. I should also mention that Heroku builds are success with a successful `npm instal`.

**Comment by ljharb** [comment]
Why would eslint ever be running in a production environment? The linter is part of tests, not part of the build process.

**Comment by ljharb** [comment]
I have no idea what craco is.

The best solution, regardless, will be to do your production builds in CI, make an artifact, and deploy that to heroku, instead of using their pre-2010 model of "git push to production to deploy".

Short of that, you'd need to somehow install dev deps in production, which would probably work with react-scripts as well.
